src/lib/csv_stats.lua

```diff
@@ -0,0 +1,81 @@
+module(..., package.seeall)
+
+local timer = require("core.timer")
+local engine = require("core.app")
+local counter = require("core.counter")
+
+CSVStatsTimer = {}
+
+-- A timer that monitors packet rate and bit rate on a set of links,
+-- printing the data out to a CSV file.
+function CSVStatsTimer:new(filename)
+   local file = filename and io.open(filename, "w") or io.stdout
+   local o = { header='Time (s)', link_data={}, file=file, period=1 }
+   return setmetatable(o, {__index = CSVStatsTimer})
+end
+
+-- Add links from an app whose identifier is ID to the CSV timer.  If
+-- present, LINKS is an array of strings identifying a subset of links
+-- to monitor.  The optional LINK_NAMES table maps link names to
+-- human-readable names, for the column headers.
+function CSVStatsTimer:add_app(id, links, link_names)
+   local function add_link_data(name, link)
+      local pretty_name = (link_names and link_names[name]) or name
+      local h = (',%s MPPS,%s Gbps'):format(pretty_name, pretty_name)
+      self.header = self.header..h
+      local data = {
+         txpackets = link.stats.txpackets,
+         txbytes = link.stats.txbytes,
+      }
+      table.insert(self.link_data, data)
+   end
+
+   local app = assert(engine.app_table[id], "App named "..id.." not found")
+   if links == nil then
```

Comment:
I tend to compare to nil if the variable is a cdata object, if it's a Lua variable ```if not links```. In the codebase there's a mixed approach, but it seems ```if not..`` is preferred. I leave it to you.